proposals/NNNN-non-escapable.md

```diff
@@ -0,0 +1,387 @@
+protocol Escapable: ~Copyable {}
+```
+
+#### In concrete contexts, `~Escapable` indicates non-escapability
+
+Using the same approach as used for `~Copyable` and `Copyable`, we use `~Escapable` to indicate the lack of the `Escapable` attribute on a type.
+
+```
+// Example: A type that is not escapable
+struct NotEscapable: ~Escapable { }
+
+// Example: Basic limits on ~Escapable types
+func f() -> NotEscapable {
+  let ne = NotEscapable()
+  borrowingFunc(ne) // OK to pass to borrowing function
+  let another = ne // OK to make local copies
+  globalVar = ne // 🛑 Cannot assign ~Escapable type to a global var
+  return ne // 🛑 Cannot return ~Escapable type
+}
+```
+
+Without a `~Escapable` marker, the default for any type is to be escapable.  Since `~Escapable` indicates the lack of a capability, you cannot put this in an extension.
+
+```
+// Example: Escapable by default
+struct Ordinary { }
+extension Ordinary: ~Escapable // 🛑 Extensions cannot remove a capability
+```
+
+Classes cannot be declared `~Escapable`.
+
+#### In generic contexts,  `~Escapable` marks the lack of an Escapable requirement
+
+When used in a generic context, `~Escapable` allows you to define functions or types that can work with values that might or might not be escapable.
+That is, `~Escapable` indicates the lack of an escapable requirement.
+Since the values might not be escapable, the compiler must conservatively prevent the values from escaping:
+
+```
+// Example: In generic contexts, ~Escapable is
+// the lack of an Escapable requirement.
+func f<MaybeEscapable: ~Escapable>(_ value: MaybeEscapable) {
+  // `value` might or might not be Escapable
+  globalVar = value // 🛑 Cannot assign possibly-non-escapable type to a global var
+}
+f(NotEscapable()) // Ok to call with non-escapable argument
+f(7) // Ok to call with escapable argument
+```
+
+This also permits the definition of types whose escapability varies depending on their generic arguments.
+As with other conditional behaviors, this is expressed by using an extension to conditionally add a new capability to the type:
+
+```
+// Example: Conditionally Escapable generic type
+// By default, Box is itself non-escapable
+struct Box<T: ~Escapable>: ~Escapable {
+  var t: T
+}
+
+// Box gains the ability to escape whenever its
+// generic argument is Escapable
+extension Box: Escapable when T: Escapable { }
+```
+
+**Note:**  There is no relationship between `Copyable` and `Escapable`.
+Copyable or non-copyable types can be escapable or non-escapable.
+
+#### Constraints on non-escapable local variables
+
+A non-escapable value can be freely copied and passed into other functions as long as the usage can guarantee that the value does not persist beyond the current scope:
+
+```
+// Example: Local variable with non-escapable type
+func borrowingFunc(_: borrowing NotEscapable) { ... }
+func consumingFunc(_: consuming NotEscapable) { ... }
+func inoutFunc(_: inout NotEscapable) { ... }
```

Comment:
It'd be good to add a case for async functions and async let here, it's barely mentioned in the proposal.

Reply:
I'll see what I can do about including async examples.

Until then, note that the comments about functions apply to _all_ functions, regardless of whether they are async or not, throwing or not.

Similarly, `async let` guarantees that the called function will return before the current function completes, so it's safe to pass a `~Escapable` value to a  function called with `async let`.